relay-event-schema/src/protocol/nel.rs

```diff
@@ -0,0 +1,251 @@
+/// infrastructure). We also use a version of this structure to deserialize from raw JSON
+/// via serde.
+///
+/// See <https://w3c.github.io/network-error-logging/>
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct NetworkReport {
+    /// The age of the report since it got collected and before it got sent.
+    pub age: Annotated<i64>,
+    /// The type of the report.
+    #[metastructure(field = "type")]
+    pub ty: Annotated<String>,
+    /// The URL of the document in which the error occurred.
+    #[metastructure(pii = "true")]
+    pub url: Annotated<String>,
+    /// The User-Agent HTTP header.
+    pub user_agent: Annotated<String>,
+    /// The body of the NEL report.
+    pub body: Annotated<Body>,
+}
+
+impl NetworkReport {
+    fn get_request(&self) -> Request {
+        let headers = match self.user_agent.value() {
+            Some(ref user_agent) if !user_agent.is_empty() => {
+                Annotated::new(Headers(PairList(vec![Annotated::new((
+                    Annotated::new(HeaderName::new("User-Agent")),
+                    Annotated::new(HeaderValue::new(user_agent)),
+                ))])))
+            }
+            Some(_) | None => Annotated::empty(),
+        };
+        Request {
+            url: self.url.clone(),
+            headers,
+            ..Request::default()
+        }
+    }
+    pub fn apply_to_event(data: &[u8], event: &mut Event) -> Result<(), ReportError> {
+        let mut nel: Annotated<NetworkReport> =
+            Annotated::from_json_bytes(data).map_err(ReportError::InvalidJson)?;
+
+        if let Some(nel) = nel.value_mut() {
+            if let Some(body) = nel.body.value() {
+                event.logentry = Annotated::new(LogEntry::from({
+                    if nel.ty.value().map_or("", |v| v.as_str()) == "http.error" {
+                        format!(
+                            "{} / {} ({})",
+                            body.phase.as_str().unwrap_or(""),
+                            body.ty.as_str().unwrap_or(""),
+                            body.status_code.value().unwrap_or(&0)
+                        )
+                    } else {
+                        format!(
+                            "{} / {}",
+                            body.phase.as_str().unwrap_or(""),
+                            body.ty.as_str().unwrap_or(""),
+                        )
+                    }
+                }));
+
+                event.request = Annotated::new(nel.get_request());
+                event.logger = Annotated::from("nel".to_string());
+
+                // Exrtact common tags.
+                let tags = event.tags.get_or_insert_with(Tags::default);
+                if let Some(ref method) = body.method.value() {
+                    tags.push(Annotated::new(TagEntry::from_pair((
```

Comment:
Please try to avoid creating tags. Instead, we have two better options:
- Put these fields into a `context`, where they will be automatically indexed by Snuba.
- Explicitly index fields from the NEL structure in case we're keeping it.

Tags are today meant for user-supplied data.